docker/service.go

```diff
@@ -186,15 +200,59 @@ func (s *Service) up(imageName string, create bool) error {
 	}
 
 	return s.eachContainer(func(c *Container) error {
-		if outOfSync, err := c.OutOfSync(); err != nil {
-			return err
-		} else if outOfSync {
-			logrus.Warnf("%s needs rebuilding", s.Name())
+		if create {
+			if err := s.rebuildIfNeeded(c); err != nil {
+				return err
+			}
 		}
+
 		return c.Up(imageName)
 	})
 }
 
+func (s *Service) rebuildIfNeeded(c *Container) error {
+	imageName, err := s.build()
+	if err != nil {
+		return err
+	}
+
+	if s.context.NoRecreate {
+		return nil
+	}
+
+	outOfSync, err := c.OutOfSync()
+	if err != nil {
+		return err
+	}
+
+	containerInfo, err := c.findInfo()
+	if containerInfo == nil || err != nil {
+		return err
+	}
+	name := containerInfo.Name[1:]
+
+	origRebuildLabel := containerInfo.Config.Labels[REBUILD.Str()]
+	newRebuildLabel := s.Config().Labels.MapParts()[REBUILD.Str()]
+	rebuildLabelChanged := newRebuildLabel != origRebuildLabel
+	logrus.WithFields(logrus.Fields{
+		"origRebuildLabel":    origRebuildLabel,
+		"newRebuildLabel":     newRebuildLabel,
+		"outOfSync":           outOfSync,
+		"ForceRecreate":       s.context.ForceRecreate,
+		"rebuildLabelChanged": rebuildLabelChanged}).Debug("Rebuild values")
+
+	if s.context.ForceRecreate || origRebuildLabel == "always" || rebuildLabelChanged || origRebuildLabel != "false" && outOfSync {
+		logrus.Infof("Rebuilding %s", name)
+		if _, err := c.Rebuild(imageName); err != nil {
+			return err
+		}
```

Comment:
I don't see where this `REBUILD` label is being set on the container.

Why does it matter what the flag was when we originally created the container? I think the only thing that matters is the current flags to `up`.